plutus-core/plutus-core/src/PlutusCore/Core/Type.hs

```diff
@@ -91,14 +98,39 @@ newtype Normalized a = Normalized
       deriving Applicative via Identity
 
 -- | All kinds of uniques an entity contains.
-type family HasUniques a :: Constraint
+type family HasUniques a :: GHC.Constraint
 type instance HasUniques (Kind ann) = ()
 type instance HasUniques (Type tyname uni ann) = HasUnique tyname TypeUnique
 type instance HasUniques (Term tyname name uni fun ann) =
     (HasUnique tyname TypeUnique, HasUnique name TermUnique)
 type instance HasUniques (Program tyname name uni fun ann) =
     HasUniques (Term tyname name uni fun ann)
 
+-- | A class for converting Haskell kinds to PLC kinds.
+class KnownKind kind where
+    knownKind :: proxy kind -> Kind ()
+
+instance KnownKind GHC.Type where
+    knownKind _ = Type ()
+
+instance (KnownKind dom, KnownKind cod) => KnownKind (dom -> cod) where
+    knownKind _ = KindArrow () (knownKind $ Proxy @dom) (knownKind $ Proxy @cod)
+
+-- We need this for type checking Plutus, however we get Plutus types/terms/programs by either
+-- producing them directly or by parsing/decoding them and in both the cases we have access to the
+-- @Typeable@ information for the Haskell kind of a type and so we could just keep it around
+-- (instead of throwing it away like we do now) and compute the Plutus kind directly from that.
+-- That might be less efficient and probably requires updating the Plutus Tx compiler, so we went
+-- with the simplest option for now and it's to have a class. Providing an instance per universe is
+-- no big deal.
+-- | For getting the Plutus kind of a type from the universe.
+class ToKind (uni :: GHC.Type -> GHC.Type) where
+    toKind :: forall k (a :: k). uni (T a) -> Kind ()
```

Comment:
Sure, do you have a suggestion?

> I was initially convinced `T` must be a type variable since that's always what one-letter types are. 

But it starts with a capital letter!

Reply:
Sure, so the "it's a type variable" and "it's an actual type" intuitions warred in my head and I was confused.

I note that if we did define `Tag`, then most uses of `T` would go away, and it would matter a *lot* less what we called it. Then we could give it a longer name and not care.

Reply:
> I was initially convinced `T` must be a type variable since that's always what one-letter types are. Can we pick a different name?

Hm, I used that "escape" term and I liked it. What do you think about `Esc`? For example:

```haskell
data DefaultUni a where
    DefaultUniBool :: DefaultUni (Esc Bool)
    DefaultUniList :: DefaultUni (Esc [])    
```

That makes it more clear that `Esc` is just syntactic noise like `\`.

Reply:
^ also pinging @kwxm